Add functionality to automatically upload logs to Vertex Tensorboard

SurbhiJainUSC opened the pull request:
- Add functionality to upload data in `config.tensorboard_dir` to Tensorboard in Vertex AI
- XPK users won't have to create a Tensorboard in Vertex AI. XPK will automatically handle that.
- For non-XPK cases, users can choose to manually create Tensorboard instance in Vertex AI cloud console or set the configuration to create instance in MaxText

Note: Uploader to upload logs to Vertex Tensorboard is only supported for `Tensorflow < 2.15.0`. Vertex AI team is working on fixing the issue to support the latest TF versions.

rwitten commented on MaxText/configs/base.yml:
```diff
@@ -247,3 +247,28 @@ decode_sampling_temperature: 1.
+
+# Vertex AI Tensorboard Configurations
+# Set to True to create Vertex Tensorboard instance in GCP
+# Set to False if running via XPK as Vertex Tensorboard and Experiment will be created by XPK
```
Can you explain in what settings you'll want the different values here? Do we always set this to true for local runs to create the Vertex Tensorboard? Or?

SurbhiJainUSC replied:
Added details about different scenarios in README.

rwitten commented on MaxText/configs/base.yml:
```diff
@@ -247,3 +247,28 @@ decode_sampling_temperature: 1.
+# Project to create Vertex Tensorboard in if create_vertex_tensorboard=True
+# Project where Vertex Tensorboard to upload data exists if upload_data_to_tensorboard=True
+# Set to blank if running via XPK as XPK will set this as an environment variable
+vertex_tensorboard_project: ""
```
Shouldn't this typically not be necessary -- it will inherit from the current project?

rwitten replied:
(If not local?)

SurbhiJainUSC replied:
Yes, added this to cater local development as well.

rwitten replied:
But isn't local development in a project? Just trying to understand.

rwitten replied:
feature sounds great but now the comment is wrong right?

SurbhiJainUSC replied:
Updated the comment.

rwitten commented on MaxText/configs/base.yml:
```diff
@@ -247,3 +247,28 @@ decode_sampling_temperature: 1.
+# Location where Vertex Tensorboard exists if upload_data_to_tensorboard=True
+# Set to blank if running via XPK as XPK will set this as an environment variable
+# Vertex AI supported locations: https://cloud.google.com/vertex-ai/docs/general/locations#available-regions
+vertex_tensorboard_location: ""
```
Do you think we could call this vertex_tensorboard_region assuming the value would be a region?

SurbhiJainUSC replied:
Sure, will update this.

rwitten commented on MaxText/configs/base.yml:
```diff
@@ -247,3 +247,28 @@ decode_sampling_temperature: 1.
+# Name of Vertex Tensorboard instance to create if create_vertex_tensorboard=True
+# Name of Vertex Tensorboard to use to upload data if upload_data_to_tensorboard=True
+# Set to blank if running via XPK as XPK will set this as an environment variable 
+vertex_tensorboard_name: ""
```
Shouldn't this be run_name regardless of XPK?

SurbhiJainUSC replied:
Hmm, let me simplify the configurations more. I think it causing a lot of confusion with multiple configs. Let me think about a good way to manage these.

rwitten commented on MaxText/configs/base.yml:
```diff
@@ -247,3 +247,28 @@ decode_sampling_temperature: 1.
+# Name of Vertex Experiment to create if create_vertex_tensorboard=True
+# Name of Vertex Experiment to use to upload data if upload_data_to_tensorboard=True
+# Set to blank if running via XPK as XPK will set this as an environment variable
+vertex_experiment_name: ""
```
I don't understand the Vertex Experiment terminology (and realistically neither will users). We need to explain it! (Like when do multiple runs go into the same experiment vs. different experiments.)

rwitten replied:
This comment and some others might be handled in the README! But then let's link to the README and try to get one source of truth.

SurbhiJainUSC replied:
Updated README with more details.

rwitten commented on MaxText/configs/base.yml:
```diff
@@ -247,3 +247,28 @@ decode_sampling_temperature: 1.
+vertex_experiment_name: ""
+# Set to True to upload data in 'config.tensorboard_dir' like profiles/scalar metrics to Vertex Tensorboard
+# Set to False if running via XPK as XPK will set this as an environment variable
+upload_data_to_tensorboard: False
```
Shouldn't this be `upload_tensorboard_data_to_vertex`?

SurbhiJainUSC replied:
Or  should we call it `upload_data_to_vertex_tensorboard`?

rwitten commented on MaxText/train.py:
```diff
@@ -480,14 +482,28 @@ def main(argv: Sequence[str]) -> None:
   validate_train_config(config)
   os.environ["TFDS_DATA_DIR"] = config.dataset_path
+
+  if config.create_vertex_tensorboard and jax.process_index()==0:
```
Let's encapsulate this in a function for readability. Happy for that function to go in train.py or elsewhere.

rwitten commented on MaxText/train.py:
```diff
@@ -480,14 +482,28 @@ def main(argv: Sequence[str]) -> None:
+      train_loop(config)
+    finally:
+      if vertex_uploader_flag:
+        uploader.stop_upload_to_tensorboard()
```
Shouldn't this be handle with a destructor on uploader?

SurbhiJainUSC replied:
Do you mean using `__del__` in python? I believe it is not guaranteed that the object will be destroyed: https://docs.python.org/3/reference/datamodel.html#object.__del__. Therefore, I have explicitly added a method to destroy the thread. Also, this design was followed to align with the Vertex AI APIs: https://cloud.google.com/vertex-ai/docs/experiments/tensorboard-upload-existing-logs. 

What are your thoughts?

rwitten replied:
I don't understand. isn't the Python3 doc saying that in a teardown event we might not call your destructor? But in that event do you want to teardown this? (Recall in general we can segfault so there is no promise.)

What is this `stop_upload_to_tensorboard` doing exactly?

SurbhiJainUSC replied:
Added a new `VertexTensorboardManager` class that will take care of destroying the thread when the program ends.

rwitten commented on README.md:
```diff
@@ -219,3 +219,53 @@ MaxText supports training and inference of various open models. Follow user guid
 * [Gemma](https://ai.google.dev/gemma): a family of open-weights Large Language Model (LLM) by [Google DeepMind](https://deepmind.google/), based on Gemini research and technology. \
 You can run decode and finetuning using instructions mentioned [here](https://github.com/google/maxtext/blob/main/getting_started/Run_Gemma.md).
+
+## Automatically Upload Logs to Vertex Tensorboard
```
Let's move this in a standalone file. I'm trying to simplify the README. We will link to the standalone file in the primary README.

https://github.com/google/maxtext/tree/main/getting_started -- place for various explainers including this one.

SurbhiJainUSC replied:
Done!

rwitten commented on README.md:
````diff
@@ -219,3 +219,53 @@ MaxText supports training and inference of various open models. Follow user guid
+```
+The above configuration will upload logs in `config.tensorboard_dir` to a manually created Vertex Tensorboard: `test-tb-instance` and Experiment: `test-experiment` in `us-central1` in `test-project`.
+
+**Scenario 3: Create Vertex Tensorboard & Upload Logs in MaxText**
````
Generally what is the use case for this scenario? (Is Vertex TB something we foresee the user setting up many times? Or is it a once per project kind of thing?)

SurbhiJainUSC replied:
Generally, users should set up Vertex TB once per project. But, one TB instance can only support around 10K Vertex Experiments. So, if for for some reason you have more than 10K Experiments attached to a Vertex TB, then you would have to create a new TB instance. This scenario takes into consideration this particular use case.

Scenario 2 is different from this one in the sense that you can also manually create the TB though Google Cloud console.

rwitten reviewed (changes requested):
Lots of questions! But looks amazing!

SurbhiJainUSC commented:
> Lots of questions! But looks amazing!

Thank you Rafi for your suggestions!
I have simplified the process of creating Vertex AI Tensorboard for testing MaxText on GCE. I have also updated the README to explain different scenarios to upload logs to Vertex Tensorboard. Let me know if this looks good.

rwitten reviewed (changes requested):
The explicit teardown is still a bit weird but looks good otherwise!

rwitten commented on MaxText/train.py:
```diff
@@ -166,6 +167,34 @@ def save_checkpoint(checkpoint_manager, step, state, dataset_type='c4', data_ite
         items=orbax.checkpoint.args.PyTreeSave(item=state)
         ))
 
+def configure_upload_to_vertex_tensorboard(config, vertex_tensorboard_manager):
```
does this feel like it should be in vertex_tensorboard.py?

rwitten replied:
just a feeling

SurbhiJainUSC replied:
Moved to VertexTensorboardManager class.

rwitten commented on README.md:
```diff
@@ -219,3 +219,6 @@ MaxText supports training and inference of various open models. Follow user guid
 You can run decode and finetuning using instructions mentioned [here](https://github.com/google/maxtext/blob/main/getting_started/Run_Gemma.md).
+
+## Automatically Upload Logs to Vertex Tensorboard
+MaxText supports automatic upload of logs collected in a directory to a Tensorboard instance in Vertex AI. Follow [user guide](https://github.com/google/maxtext/tree/main/getting_started/Use_Vertex_AI_Tensorboard.md) to know more.
```
nit: should be `[user guide](getting_started/Use_Vertex_AI_Tensorboard.md)` or similar. (That way when folks make a change their thing updates. (I'm trying to move other absolute links to be relative links.)

rwitten reviewed (approved):
Approved (but there are some nits I want you to cleanup)
